### ISSUE TITLE
Incompatibility with Mockery

### ISSUE THREAD

**Issue by matzeeable** [antecedent/patchwork#99]
Hi!

I am using PHPUnit 7 and want to unit test a WordPress plugin (I can not go with PHPUnit 8 because WordPress has official an incompatibility, see [here](https://core.trac.wordpress.org/ticket/46149)). Additionally I use [mockery](https://github.com/mockery/mockery/) as mock framework instead of PHPUnit's mock builder.

I have the following strange situation, my test file `MyAwesomeTest.php` looks like this

```php
<?php
declare(strict_types=1);

use Mockery\Adapter\Phpunit\MockeryPHPUnitIntegration;
use Mockery\MockInterface;
use PHPUnit\Framework\TestCase;

// Bootstrap composer
require_once __DIR__ . '/../../vendor/antecedent/patchwork/Patchwork.php';
require_once __DIR__ . '/../../vendor/autoload.php';

trait Foo {
    public function foo() {
        return $this->anotherFoo();
    }

    public function anotherFoo() {
        return 'my value';
    }
}

class Bar {
    use Foo;
}

class MyAwesomeTest extends TestCase {
    use MockeryPHPUnitIntegration;

    public function testFoo() {
        /** @var MockInterface|Bar */
        $mock = Mockery::mock(Bar::class);
        $mock
            ->makePartial()
            ->shouldReceive('anotherFoo')
            ->once()
            ->andReturns('another value');

        $this->assertEquals('another value', $mock->foo());
    }
}
```

Running the test gives the following result:

```bash
matzeeable@MSI:$ vendor/bin/phpunit test/php/MyAwesomeTest.php 
PHPUnit 7.5.18-3-g6aab04011 by Sebastian Bergmann and contributors.

E                                                                   1 / 1 (100%)

Time: 528 ms, Memory: 12.00 MB

There was 1 error:

1) MyAwesomeTest::testFoo
ReflectionException: Trying to invoke non static method Bar::foo() without an object

[...]/vendor/antecedent/patchwork/src/CallRerouting.php:526
[...]/vendor/antecedent/patchwork/src/CallRerouting.php:353
[...]/vendor/antecedent/patchwork/src/CallRerouting.php:535
[...]/vendor/antecedent/patchwork/src/CallRerouting.php:285
[...]/vendor/antecedent/patchwork/src/CallRerouting.php:317
[...]/vendor/antecedent/patchwork/src/Stack.php:27
[...]/vendor/antecedent/patchwork/src/CallRerouting.php:328
[...]/test/php/MyAwesomeTest.php:38

ERRORS!
Tests: 1, Assertions: 0, Errors: 1.
```

But if I do the following (comment out Patchwork):

```php
// require_once __DIR__ . '/../../vendor/antecedent/patchwork/Patchwork.php';
```

I get correct results:

```bash
matzeeable@MSI:$ vendor/bin/phpunit test/php/MyAwesomeTest.php 
PHPUnit 7.5.18-3-g6aab04011 by Sebastian Bergmann and contributors.

.                                                                   1 / 1 (100%)

Time: 272 ms, Memory: 4.00 MB

OK (1 test, 2 assertions)
```

Is there any incompatibility with Mockery + Patchwork?

Versions:

|Name|Version|
|--|--|
|PHPUnit|7.5.18-3-g6aab04011|
|PHP|7.2.17-0ubuntu0.18.04.1 (cli) (built: Apr 18 2019 14:12:38) ( NTS )|
|Mockery|dev-master|
|Patchwork|2.1.11|

Regards, Matthew 🙂

**Comment by antecedent** [comment]
Hi, do you happen to have a `patchwork.json` in your setup by any chance?

**Comment by matzeeable** [comment]
Thanks for your fast response! Yeah, my `patchwork.json` looks like this:

```json
{
    "redefinable-internals": ["call_user_func", "error_log"]
}
```

I just figured out if I remove the internals and make the `packwork.json` like the below, the test passes:

```json
{
    "redefinable-internals": []
}
```

Do you need anything else to know?

**Comment by antecedent** [comment]
I reproduced it with Mockery 1.3.0 and PHPUnit 8.5.0. A fix is now in the `master` branch; please try it.

**Comment by matzeeable** [comment]
I implemented the fix on my local `vendor` folder and the error is gone, thank you!

Unfortunately I can not rely on the `dev-master` because I use all the packages together with [WP_Mock](https://github.com/10up/wp_mock/blob/master/composer.json#L10). When do you release a new version? 🙂

**Comment by matzeeable** [comment]
Okay, I just found another strange thing I think you can reprocedure... Just try the following test case:

```php
class Service {
    private $foo;

    public function __construct($foo) {
        $this->foo = $foo;
    }
}

trait Foo {
    private $service;

    public function foo() {
        $this->service = new Service($this);
        $this->getService();
        // Do something here with service ...
        return $this->anotherFoo();
    }

    public function getService() {
        return $this->service;
    }

    public function anotherFoo() {
        return 'my value';
    }
}

class Bar {
    use Foo;
}
```

and the test case keeps the same. Again here, if I do not load patchwork it work as expected. If I have patchwork active, the following error occurs:

```bash
1) MyAwesomeTest::testFoo
call_user_func_array() expects parameter 1 to be a valid callback, cannot access parent:: when no class scope is active

XXX/test/php/MyAwesomeTest.php:21
XXX/vendor/antecedent/patchwork/src/CallRerouting.php:529
XXX/vendor/antecedent/patchwork/src/CallRerouting.php:353
XXX/vendor/antecedent/patchwork/src/CallRerouting.php:538
XXX/vendor/antecedent/patchwork/src/CallRerouting.php:285
XXX/vendor/antecedent/patchwork/src/CallRerouting.php:317
XXX/vendor/antecedent/patchwork/src/Stack.php:27
XXX/vendor/antecedent/patchwork/src/CallRerouting.php:328
XXX/test/php/MyAwesomeTest.php:55
```

**Comment by antecedent** [comment]
This took considerably longer, but should now be working. Could you confirm that?

I would release a new version shortly thereafter if everything works as intended.

**Comment by matzeeable** [comment]
It works as expected! Should I let you know if I find further compatibility issues with `patchwork` + `mockery`? 🙂

**Comment by antecedent** [comment]
Great! The fixes are now in version 2.1.12.

> Should I let you know if I find further compatibility issues with `patchwork` + `mockery`? 🙂

Yes, please do :)
